src/main/kotlin/eu/europa/ec/eudi/pidissuer/adapter/out/jose/DecryptCNonceWithNimbus.kt

```diff
@@ -0,0 +1,95 @@
+import com.nimbusds.jose.jwk.RSAKey
+import com.nimbusds.jose.jwk.source.ImmutableJWKSet
+import com.nimbusds.jose.proc.*
+import com.nimbusds.jwt.EncryptedJWT
+import com.nimbusds.jwt.JWTClaimsSet
+import com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier
+import com.nimbusds.jwt.proc.DefaultJWTProcessor
+import eu.europa.ec.eudi.pidissuer.adapter.out.IssuerSigningKey
+import eu.europa.ec.eudi.pidissuer.adapter.out.signingAlgorithm
+import eu.europa.ec.eudi.pidissuer.domain.CNonce
+import eu.europa.ec.eudi.pidissuer.domain.CredentialIssuerId
+import eu.europa.ec.eudi.pidissuer.port.out.jose.DecryptCNonce
+import org.bouncycastle.jce.provider.BouncyCastleProvider
+import java.time.Duration
+
+/**
+ * Nimbus implementation of [DecryptCNonce] that decrypts an [EncryptedJWT] as a [CNonce].
+ */
+internal class DecryptCNonceWithNimbus(
+    private val issuer: CredentialIssuerId,
+    private val decryptionKey: RSAKey,
+    private val signingKey: IssuerSigningKey,
+) : DecryptCNonce {
+    init {
+        require(decryptionKey.isPrivate) { "a private key is required for decryption" }
+    }
+
+    private val processor = DefaultJWTProcessor<SecurityContext>()
+        .apply {
+            val jcaProvider = BouncyCastleProvider()
+            val typeVerifier = DefaultJOSEObjectTypeVerifier<SecurityContext>(JOSEObjectType("cnonce+jwt"))
+
+            jweTypeVerifier = typeVerifier
+            jweKeySelector = JWEDecryptionKeySelector(
+                JWEAlgorithm.RSA_OAEP_512,
+                EncryptionMethod.XC20P,
+                ImmutableJWKSet(JWKSet(decryptionKey)),
+            )
+            jweDecrypterFactory = DefaultJWEDecrypterFactory().apply {
+                jcaContext.provider = jcaProvider
+            }
+
+            jwsTypeVerifier = typeVerifier
+            jwsKeySelector = JWSVerificationKeySelector(
+                signingKey.signingAlgorithm,
+                ImmutableJWKSet(JWKSet(signingKey.key)),
+            )
+            jwsVerifierFactory = DefaultJWSVerifierFactory().apply {
+                jcaContext.provider = jcaProvider
+            }
+
+            jwtClaimsSetVerifier = DefaultJWTClaimsVerifier(
+                issuer.externalForm,
+                JWTClaimsSet.Builder()
+                    .issuer(issuer.externalForm)
+                    .audience(issuer.externalForm)
+                    .build(),
+                setOf("iss", "aud", "cnonce", "exi", "iat", "exp"),
+            )
+        }
+
+    override suspend fun invoke(encrypted: String): Result<CNonce> = result {
+        val jwt = EncryptedJWT.parse(encrypted)
+        val claimSet = processor.process(jwt, null)
+        CNonce(
+            requireNotNull(claimSet.getStringClaim("cnonce")) { "missing 'cnonce' claim" },
+            requireNotNull(claimSet.issueTime.toInstant()) { "missing 'iat' claim" },
+            requireNotNull(claimSet.getLongClaim("exi")?.let { Duration.ofSeconds(it) }) { "missing 'exi' claim" },
```

Comment:
Given that there is an `exp` claim, what is the use of `exi`?

Reply:
`exi` is the IANA registered claim for `Expires In`.
For CNonce we need `iat`, `cnonce` and `exi`. `exp` is basically `iat` + `exi`.

https://www.iana.org/assignments/jwt/jwt.xhtml